src/transforms/throttle.rs

```diff
@@ -40,6 +56,17 @@ pub struct ThrottleConfig {
 
     /// A logical condition used to exclude events from sampling.
     exclude: Option<AnyCondition>,
+
+    /// The throttling unit to use.
+    #[serde(default = "default_unit")]
+    unit: ThrottleUnit,
+
+    /// The encoding to use if throttling by bytes
+    encoding: Option<EncodingConfig>,
```

Comment:
I think this could be done with an `enum` using serde's `internally_tagged` mode, which would make it impossible to configure the "bytes" mode without an encoding:
```rust
enum ThrottleMode {
    Events,
    Bytes {
        encoding: EncodingConfig,
    },
}
```

Reply:
From what I understand, this is possible, with a caveat. If we want these fields at the top-level, then I believe we need to use the `flatten` serde config. However, this cannot be used in conjunction with `deny_unknown_fields`. Removing `deny_unknown_fields` seems to spoil the certainty that this change is supposed to introduce. It is very possible I am missing something though.

Alternatively, we can avoid flattening. Nothing wrong with this, but the name gets a little tricky because we may end up with structs that look like this. I can try to experiment with options, but any further suggestions are appreciated. I am learning as I am going.

```
mode:
  mode: events
```

Reply:
Yes, serde enum encoding is tricky and, yes, dropping `deny_unknown_fields` is undesirable. I don't have a good suggestion here other than a custom deserializer, so probably leaving the encoding as a separate parameter is the most reasonable path forward.

Reply:
I happened to come across a similar scenario with the syslog source. It has a `mode` field which is an enum with a lot of configuration options. It uses flattening and internal tagging while accepting unknown fields, just as this PR's current state is doing. I think this is a sign that this is the preferred approach.
https://github.com/vectordotdev/vector/blob/master/src/sources/syslog.rs#L36-L37